File: Python/matmat/matmat_jik.py
# matmat_jik.py
#
# Daniel R. Reynolds
# SMU Mathematics
# Math 5316
# Spring 2019

def matmat_jik(A,X):
    """
    Usage: B = matmat_jik(A,X)

    Function to perform matrix-matrix multiplication with "jik" loop ordering.

    Inputs: A is a matrix (m x n numpy matrix)
            X is a matrix (n x p numpy matrix)
    Output: B is a matrix (m x p numpy matrix)
    """

    # imports
    import numpy as np

    # get problem dimensions
    m, n = np.shape(A)
    n2, p = np.shape(X)

    # check that A and X are compatible
    if (n != n2):
        raise ValueError("matmat_jik error: A and X are incompatible")

    # initialize output
    B = np.zeros([m,p], dtype=float)

    # perform product
    for j in range(p):
        for i in range(m):
            B[i,j] += A[i,:] @ X[:,j]

    return B

File: Matlab/matmat/matmat_jik.m
function B = matmat_jik(A,X)
% Usage: B = matmat_jik(A,X)
%
% Function to perform matrix-matrix multiplication with "jik" loop ordering.
%
% Inputs: A is a matrix (m x n numpy matrix)
%         X is a matrix (n x p numpy matrix)
% Output: B is a matrix (m x p numpy matrix)
%
% Daniel R. Reynolds
% SMU Mathematics
% Math 5316
% Spring 2019

% get problem dimensions
[m,n]  = size(A);
[n2,p] = size(X);

% check that A and X are compatible
if (n ~= n2)
   error('matmat_jik error: A and X are incompatible')
end

% initialize output
B = zeros(m,p);

% perform product
for j=1:p
   for i=1:m
      for k=1:n
         B(i,j) = B(i,j) + A(i,k)*X(k,j);
      end
   end
end
